src/lucky_router/matcher.cr

```diff
@@ -30,25 +31,33 @@ class LuckyRouter::Matcher(T)
 
     path_without_optional_params = all_path_parts.reject(&.optional?)
 
-    process_and_add_path(method, path_without_optional_params, payload)
+    process_and_add_path(method, path_without_optional_params, payload, path)
     optional_parts.each do |optional_part|
       path_without_optional_params << optional_part
-      process_and_add_path(method, path_without_optional_params, payload)
+      process_and_add_path(method, path_without_optional_params, payload, path)
     end
     if glob_part
       path_without_optional_params << glob_part
-      process_and_add_path(method, path_without_optional_params, payload)
+      process_and_add_path(method, path_without_optional_params, payload, path)
     end
   end
 
-  private def process_and_add_path(method : String, parts : Array(PathPart), payload : T)
+  private def process_and_add_path(method : String, parts : Array(PathPart), payload : T, path : String)
     if method.downcase == "get"
       root.process_parts(parts, "head", payload)
     end
 
+    duplicate_check(method, parts, path)
```

Comment:
This couldn't be added to the `validate!` method because it runs before all of the optional and glob logic happens to create all the separate paths that also need to be duplicate checked.